### PR TITLE
Jan kubalek/component versioning

### PR DESCRIPTION
**Pull request by koudis**
## Summary by CodeRabbit

* **New Features**
  * Added support for local file URIs (`file://`) in dependency downloads with strict validation.
  * Introduced Git URI transformation utilities for HTTP(S) and SSH formats.
  * Enhanced test resource management for local file URI testing.
  * Added cache variable management functions for safer testing.
  * Added argument parsing cleanup macro to manage internal variables.
  * Added comprehensive release process documentation.

* **Bug Fixes**
  * Improved error messages and diagnostics for unreachable URIs and URI type mismatches.
  * Strengthened URI validation in dependency downloads.
  * Corrected multiple typographical errors in comments, error messages, and documentation.

* **Documentation**
  * Clarified and expanded README files and installation instructions.
  * Improved documentation for argument parsing and dependency management.
  * Added comprehensive test suite READMEs and usage notes.

* **Tests**
  * Expanded and modularized test suite covering new URI types, error cases, and argument parsing.
  * Added tests for file, HTTP, Git downloads including branch and revision-specific cases.
  * Added failure tests for unreachable URIs and URI type mismatches.
  * Added cleanup and memory management tests for argument parsing.
  * Simplified and reorganized test orchestration with external test scripts.
  * Introduced detailed tests for Git archive functionality with optional disabling.
  * Added new tests for dependency cache control with improved URI handling.
  * Added tests for argument parsing edge cases, complex scenarios, and error conditions.

* **Refactor**
  * Centralized URI construction and transformation in tests.
  * Standardized dependency cache control tests with improved error message templating.
  * Reorganized argument parsing tests from inline to modular external scripts.
  * Improved test setup to skip certain tests in script mode for efficiency.

### CODE REVIEW

File: test/DEPENDENCY/download_check/fail/uri_type_mismatches/http_with_git/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+FIND_PACKAGE(CMLIB)
+
+SET(HTTP_URI "https://example.com/file.txt")
+
+CMLIB_DEPENDENCY(
+    TYPE FILE
+    URI "${HTTP_URI}"
+    URI_TYPE GIT
+    OUTPUT_PATH_VAR output_path
+)
```
[the remark this answers is not in the document]
Comment (koudis) [reply test/DEPENDENCY/download_check/fail/uri_type_mismatches/http_with_git/CMakeLists.txt]: Done. Differently, same outcome.

File: test/FILE_DOWNLOAD/CMakeLists.txt [hunk 2]
```diff
@@ -4,16 +4,36 @@ IF(NOT  DEFINED CMAKE_SCRIPT_MODE_FILE)
 	PROJECT(CMLIB_CACHE_TEST)
 ENDIF()
 
+SET(_CMLIB_TEST_GIT_ARCHIVE ON
+	CACHE BOOL
+	"Test git archive functionality. Set to OFF if you do not have SSH key added to the test repository."
+)
```
[the remark this answers is not in the document]
Comment (koudis) [reply test/FILE_DOWNLOAD/CMakeLists.txt:10]: It works with the current setup. I have added the key....

File: test/PARSE_ARGUMENTS/test_cases/basic_functionality/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	CMLIB_PARSE_ARGUMENTS(
+		PREFIX ""
+		ONE_VALUE TEST_VALUE3
+		REQUIRED TEST_VALUE3
+		P_ARGN
+			TEST_VALUE3 "empty_prefix_test"
+	)
+
+	TEST_VAR_DEFINED(__TEST_VALUE3)
+	TEST_VAR_EQUALS_LITERAL(__TEST_VALUE3 "empty_prefix_test")
```
[the remark this answers is not in the document]
Comment (koudis) [reply test/PARSE_ARGUMENTS/test_cases/basic_functionality/CMakeLists.txt:49]: Thats ok.

File: test/test_resources/test_resources.cmake [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+##
+#
+# Helper function to get file:// URI for test files from downloaded repository
+#
+# <function>(
+#     <relative_path>  // Path relative to test repository root
+#     <output_var>     // Variable to store the file:// URI
+# )
+#
+FUNCTION(TEST_RESROUCES_GET_FILE_URI relative_path output_var)
```
[the remark this answers is not in the document]
Comment (koudis) [reply test/test_resources/test_resources.cmake:96]: Done.

File: test/DEPENDENCY/download_check/fail/unreachable_uris/file_uri_unreachable/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+LIST(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}/../../../../../")
+FIND_PACKAGE(CMLIB)
+
+SET(UNREACHABLE_FILE_URI "file:///tmp/nonexistent_directory_12345/nonexistent_file.txt")
+
+CMLIB_DEPENDENCY(
+    TYPE FILE
+    URI "${UNREACHABLE_FILE_URI}"
+    OUTPUT_PATH_VAR output_path
+)
```
[the remark this answers is not in the document]
Comment (koudis) [reply test/DEPENDENCY/download_check/fail/unreachable_uris/file_uri_unreachable/CMakeLists.txt:21]: thATS FINE. The file is not reachable and the probability that someone create the same file is almoist zero.

File: README.md [hunk 6]
```diff
@@ -49,22 +48,29 @@ Full example can be found at [example/DEPENDENCY/boost_example]
...
+git clone https://github.com/cmakelib/cmakelib.git
+```
+
+To make CMLIB works add following lines to your .bashrc
+
+```bash
+echo "export CMLIB_DIR=$(PWD)/cmakelib" >> .bashrc
+# If one, global cache directory is required setup
+#echo "export CMLIB_REQUIRED_ENV_TMP_PATH=$(PWD)/cmakelib_cache" >> .bashrc
+```
```
[the remark this answers is not in the document]
Comment (koudis) [reply README.md:71]: Are you sure with the $PWD?

File: .github/workflows/examples.yml [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+      image: ghcr.io/cmakelib/test_ubuntu:2404
+    steps:
+      - name: Install SSH key
+        uses: shimataro/ssh-key-action@v2
+        with:
+          key: ${{ secrets.SSHKEY }}
+          name: id_rsa
+          known_hosts: ${{ secrets.KNOWN_HOSTS }}
+          if_key_exists: fail
+
```
[the remark this answers is not in the document]
Comment (koudis) [reply .github/workflows/examples.yml:36]: No.

File: .github/workflows/examples.yml [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+  workflow_dispatch:
+
+jobs:
+  # Wait for tests from the Tests workflow to complete
+  wait-for-tests:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - name: Wait for Tests workflow to complete
+        uses: actions/github-script@v6
```
[the remark this answers is not in the document]
Comment (koudis) [reply .github/workflows/examples.yml]: Done.

File: system_modules/CMLIB_FILE_DOWNLOAD.cmake [hunk 9]
```diff
@@ -418,10 +444,49 @@ ENDFUNCTION()
...
+			MESSAGE(FATAL_ERROR "URI '${uri}' is not valid for FILE URI_TYPE!")
+		ENDIF()
+
+		STRING(REGEX REPLACE "^file://(.+)$" "\\1" file_path "${uri}")
+		IF(NOT IS_ABSOLUTE "${file_path}")
+			MESSAGE(FATAL_ERROR "Path specified by URI must be absolute for FILE URI_TYPE. Relative path '${uri}' provided.")
+		ENDIF()
+		IF(IS_DIRECTORY "${file_path}")
+			MESSAGE(FATAL_ERROR "Path specified by URI must be file for FILE URI_TYPE. Directory '${uri}' provided.")
+		ENDIF()
```
[the remark this answers is not in the document]
Comment (koudis) [reply system_modules/CMLIB_FILE_DOWNLOAD.cmake:481]: No, not needed.
